src/NavierStokes2D.hpp

```diff
@@ -0,0 +1,633 @@
+    {
+      for (unsigned int i = 0; i < dim - 1; ++i)
+        values[i] = 0.0;
+
+      values[dim - 1] = -g;
+    }
+
+    virtual double
+    value(const Point<dim> & /*p*/,
+          const unsigned int component = 0) const override
+    {
+      if (component == dim - 1)
+        return -g;
+      else
+        return 0.0;
+    }
+
+  protected:
+    const double g = 0.0;
+  };
+
+  // Dirichlet boundary conditions.
+  class FunctionG : public Function<dim>
+  {
+  public:
+    // Constructor.
+    FunctionG() : Function<dim>(dim + 1)
+    {
+    }
+
+    virtual void
+    vector_value(const Point<dim> & /*p*/, Vector<double> &values) const override
+    {
+      values[0] = 0.;
+      values[1] = 0.;
+      values[2] = 0.;
+    }
+
+    virtual double
+    value(const Point<dim> & /*p*/, const unsigned int /*component*/) const override
+    {
+      return 0.;
+    }
+  };
+
+  // Neumann boundary conditions.
+  class FunctionH : public Function<dim>
+  {
+  public:
+    // Constructor.
+    FunctionH()
+    {
+    }
+
+    virtual double
+    value(const Point<dim> & /*p*/, const unsigned int /*component*/) const override
+    {
+      return 0.;
+    }
+  };
+
+  // Function for the initial condition.
+  class FunctionU0 : public Function<dim>
+  {
+  public:
+    FunctionU0()
+    {
+    }
+
+    virtual double
+    value(const Point<dim> & /*p*/, const unsigned int component) const
+    {
+      if (component == 0)
+        return 0.;
+      else
+        return 0.;
+    }
+    virtual void
+    vector_value(const Point<dim> & /*p*/, Vector<double> &values) const override
+    {
+      values[0] = 0.;
+      values[1] = 0.;
+      values[2] = 0.;
+    }
+  };
+
+  // Function for inlet velocity. This actually returns an object with four
+  // components (one for each velocity component, and one for the pressure), but
+  // then only the first three are really used (see the component mask when
+  // applying boundary conditions at the end of assembly). If we only return
+  // three components, however, we may get an error message due to this function
+  // being incompatible with the finite element space.
+  class InletVelocity : public Function<dim>
+  {
+  public:
+    InletVelocity()
+        : Function<dim>(dim + 1)
+    {
+    }
+
+    virtual void
+    vector_value(const Point<dim> &p, Vector<double> &values) const override
+    {
+      values[0] = 4 * u_m * p[1] * (H - p[1]) * std::sin(M_PI * get_time() / 8.) / (H * H);
+      for (unsigned int i = 1; i < dim + 1; ++i)
+        values[i] = 0.0;
+    }
+
+    virtual double
+    value(const Point<dim> &p, const unsigned int component = 0) const override
+    {
+      if (component == 0)
+        return 4 * u_m * p[1] * (H - p[1]) / (H * H);
+      else
+        return 0;
+    }
+
+    double getMeanVelocity() const
+    {
+
+      return 8 * u_m * H / 2 * H / 2 / (3.0 * H * H);
+    }
+
+  protected:
+    double H = 0.41;
+    double u_m = 0.3; // 0.3;
```

Comment:
I assume these two values are not expected to change over time. You should make them const:
```suggestion
    const double H = 0.41;
    const double u_m = 0.3; // 0.3;
```
Moreover, these are actually compile-time constants. It is probably better to declare them as such:
```suggestion
    static inline constexpr double H = 0.41;
    static inline constexpr  u_m = 0.3; // 0.3;
```
This will help the compiler in optimizing these variables out whenever it can.